### ISSUE TITLE
phi-2 training tokenizer bug when using sharegpt format

### ISSUE THREAD

**Issue by l3utterfly** [axolotl-ai-cloud/axolotl#1357]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

When tuning with sharegpt, each conversation ending should have the eos token applied, which for phi-2 is <|endoftext|>

### Current behaviour

When tokenizing a sharegpt jsonl format for phi-2 models, the "</s>" is added as a string, THEN the "<|endoftext|>" token is added afterwards, resulting in the tuned model to output BOTH "</s>" and "<|endoftext|>" when conversation turn ends.

### Steps to reproduce

1. install axolotl
2. use example phi-2 tune script
3. use sharegpt dataset with vicuna_v1.1 format

### Config yaml

```yaml
base_model: /home/layla/src/text-generation-webui/models/phi-2
model_type: AutoModelForCausalLM
tokenizer_type: AutoTokenizer

load_in_8bit: false
load_in_4bit: false
strict: false

datasets:
  - path: /home/layla/src/Layla-datasets/datasets_formatted/base/openhermes.jsonl
    ds_type: json # see other options below
    type: sharegpt
    conversation: vicuna_v1.1

dataset_prepared_path: last_run_prepared
val_set_size: 0.01
output_dir: ./out

sequence_len: 2048
sample_packing: true
pad_to_sequence_len: true

wandb_project:
wandb_entity:
wandb_watch:
wandb_run_id:
wandb_log_model:

gradient_accumulation_steps: 4
micro_batch_size: 2
num_epochs: 1
optimizer: adamw_bnb_8bit
lr_scheduler: cosine
learning_rate: 0.0000005

train_on_inputs: false
group_by_length: false
bf16: auto
fp16:
tf32: true

gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: True
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_ratio: 0.05
eval_steps: 0.1
eval_sample_packing: true
save_steps: 300
debug:
deepspeed: /home/layla/src/Layla-datasets/axolotl/configs/deepspeed/zero2.json # multi-gpu only
weight_decay: 0.0
fsdp:
fsdp_config:
resize_token_embeddings_to_32x: true
special_tokens:
  bos_token: "<|endoftext|>"
  eos_token: "<|endoftext|>"
  unk_token: "<|endoftext|>"
  pad_token: "<|endoftext|>"
```

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

main

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by l3utterfly** [comment]
Digging deeper into the code, this is where it went wrong:

`/axolotl/src/axolotl/prompters.py:325`

The `Conversation` object from fast chat has hard coded `sep2="</s>"`

Temporary workaround for phi, need to add the line: `conv.sep2 = '<|endoftext|>'` force it to phi's separator

I'm not too familiar with the rest of the codebase to fix this permanently.

**Comment by l3utterfly** [comment]
This line also needs to be added: `self._conversation.sep2 = '<|endoftext|>'`

in `/axolotl/src/axolotl/prompters.py:286`

**Comment by PhilipMay** [comment]
I also want to train phi-2 models. Is this bug still not fixed?
@l3utterfly @winglian

**Comment by mhenrichsen** [comment]
Feel free to add a PR to fix. Closing as stale for now as no one should be training phi-2 anymore.
